data/blog/enjoyable-database-testing.md

```diff
@@ -0,0 +1,277 @@
+
+```haskell
+aroundAll withSetup $ describe "list/add/delete" $ parallel $ do
+  it "return [] initially" $ withPool $ abort $ \conn -> do
+    list conn `shouldReturn` []
+  it "returns the created elements" $ withPool $ abort $ \conn -> do
+    theId <- add conn 1
+    list conn `shouldReturn` [theId]
+  it "deletes what is created" $ withPool $ abort $ \conn -> do
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` []
+```
+
+Unfortunately if we run our tests again we will notice failures.
+
+The problem is our `abort` function is wrapping everything in a `READ_COMMITTED` isolation level. This is equivalent to each statement receiving it's own snapshot of the database but we need the entire transaction to have a single consistent snapshot of the database so we need to use either `REPEATABLE_READ` or `SERIALIZABLE` isolation.
+
+In our particular example the solution is to use the higher isolation and retry on serializable errors. However this is not always possible.
+
+For instance some postgres statements, like `SKIP LOCKED`, do not run performantly in more consistent isolation levels because they provide an intrisitically inconsistent picture of the database.
+
+This made using `SERIALIZABLE` unusable in testing `postgresql-simple-queue` however I was still able utilize `parallel` to improve performance, while maintaining isolation through separate `postgres` instances.
+
+The startup cost of `tmp-postgres` is around 250 ms on Mac or 90 ms on Linux so your tests will need to be atleast 0.6 seconds but probabaly closer to 2.0 seconds for this approach to be meaningfully helpful.
+
+Here is what it would look like in our example:
+
+```haskell
+describe "list/add/delete" $ parallel $ do
+  aroundAll withSetup $ do
+    it "return [] initially" $ withPool $ abort $ \conn -> do
+      list conn `shouldReturn` []
+    it "returns the created elements" $ withPool $ abort $ \conn -> do
+      theId <- add conn 1
+      list conn `shouldReturn` [theId]
+
+  aroundAll withSetup $ do
+    it "deletes what is created" $ withPool $ abort $ \conn -> do
+      theId <- add conn 2
+      delete conn theId
+      list conn `shouldReturn` []
+```
+
+Starting a separate `postgres` instance is a big hammer. It is a heavyweight operation but can surprisingly help performance in some situations and provides the highest level of isolation. However typically `abort` is optimal.
+
+## Reuse setup with `rollback`
+
+`abort` is great for rolling back all the changes of test but sometimes we would like to only rollback some changes in a test.
+
+This situation arises when we would like to reuse some setup to run a few different assertions.
+
+```haskell
+it "returns [] after all removal operations" $ withPool $ abort $ \conn -> do
+  complexSetup conn
+  forM_ allRemovalOperations $ \op -> rollback $ op conn >> list `shouldReturn` []
+```
+
+Unlike `abort` `rollback` can be nested and does not abort the entire transaction.
```

Comment:
Unlike `abort`**,** `rollback` can be nested and does not abort the entire transaction.